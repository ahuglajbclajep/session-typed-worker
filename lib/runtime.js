async function init(selfOrWorkers) {
    let ports;
    if (selfOrWorkers === self) {
        ports = await new Promise((resolve) => (selfOrWorkers.onmessage = (e) => resolve(e.data)));
    }
    else {
        const workers = selfOrWorkers;
        const selves = Object.entries(workers).filter(([, w]) => w === self);
        if (selves.length !== 1)
            return;
        const roleOfSelf = selves[0][0];
        let portMap = Object.fromEntries(Object.keys(workers).map((r) => [r, {}]));
        for (const k of Object.keys(workers)) {
            for (const l of Object.keys(workers)) {
                if (k === l)
                    continue;
                if (!portMap[k][l]) {
                    const { port1, port2 } = new MessageChannel();
                    portMap[k][l] = port1;
                    portMap[l][k] = port2;
                }
            }
        }
        Object.entries(workers)
            .filter(([, w]) => w !== self)
            .forEach(([r, w]) => w.postMessage(portMap[r], Object.values(portMap[r])));
        ports = portMap[roleOfSelf];
    }
    const queues = Object.fromEntries(Object.keys(ports).map((r) => [r, []]));
    const context = { ports, queues };
    Object.entries(ports).forEach(([r, p]) => (p.onmessage = (e) => context.queues[r].push(e.data)));
    return context;
}
function send(channel, role, label, value) {
    channel.ports[role].postMessage([label, value]);
    return channel;
}
async function recv(channel, role) {
    const { queues, ports } = channel;
    if (0 < queues[role].length)
        return queues[role].shift();
    const onmessage = ports[role].onmessage;
    return new Promise((r) => (ports[role].onmessage = (e) => {
        ports[role].onmessage = onmessage;
        r(e.data);
    }));
}
function close(channel) { }
export { init, send, recv, close };
