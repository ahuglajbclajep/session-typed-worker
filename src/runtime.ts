import type {
  Local,
  Select,
  SelectLocals,
  Offer,
  OfferLocals,
  Close,
} from "./mpst";

type Message = [string, unknown];
type Ports = Record<string, MessagePort>;
type Context = { ports: Ports; queues: Record<string, Message[]> };

type Workers = Record<string, Window | Worker>;
type SelfOrWorkers = DedicatedWorkerGlobalScope | Workers;

async function init(selfOrWorkers: SelfOrWorkers): Promise<Local> {
  let ports: Ports;
  if (selfOrWorkers === self) {
    if (self.document)
      throw Error("The window is passed directly to the init function.");
    ports = await new Promise(
      (resolve) => (selfOrWorkers.onmessage = (e) => resolve(e.data))
    );
  } else {
    const workers = selfOrWorkers as Workers;
    const selves = Object.entries(workers).filter(([, w]) => w === self);
    if (selves.length !== 1)
      throw Error("Only one self can be passed to the init function.");
    const roleOfSelf = selves[0][0];

    let portMap: Record<string, Ports> = Object.fromEntries(
      Object.keys(workers).map((r) => [r, {}])
    );
    for (const k of Object.keys(workers)) {
      for (const l of Object.keys(workers)) {
        if (k === l) continue;
        if (!portMap[k][l]) {
          const { port1, port2 } = new MessageChannel();
          portMap[k][l] = port1;
          portMap[l][k] = port2;
        }
      }
    }
    Object.entries(workers)
      .filter(([, w]) => w !== self)
      .forEach(([r, w]) =>
        (w as Worker).postMessage(portMap[r], Object.values(portMap[r]))
      );
    ports = portMap[roleOfSelf];
  }

  const queues = Object.fromEntries(Object.keys(ports).map((r) => [r, []]));
  const context: Context = { ports, queues };
  Object.entries(ports).forEach(
    ([r, p]) => (p.onmessage = (e) => context.queues[r].push(e.data))
  );
  return context as any;
}

function send<R extends string, LS extends SelectLocals, L extends keyof LS>(
  port: Select<R, LS>,
  role: R,
  label: L,
  value: Parameters<LS[L]>[0]
): ReturnType<LS[L]> {
  ((port as any) as Context).ports[role].postMessage([label, value]);
  return port as any;
}

async function recv<R extends string, LS extends OfferLocals>(
  port: Offer<R, LS>,
  role: R
): Promise<LS> {
  const { queues, ports } = (port as any) as Context;
  if (0 < queues[role].length) {
    const [label, value] = queues[role].shift()!;
    return [label, [value, port]] as any;
  }

  const onmessage = ports[role].onmessage;
  return new Promise(
    (resolve) =>
      (ports[role].onmessage = (e) => {
        ports[role].onmessage = onmessage;
        const [label, value] = e.data as Message;
        resolve([label, [value, port]] as any);
      })
  );
}

function close(channel: Close) {}

export { init, send, recv, close };
