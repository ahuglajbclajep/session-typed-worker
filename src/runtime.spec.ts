import { send, receive } from "./runtime";
import type { Select, Offer, Close } from "./mpst";

///////////////////////////// test cases of send /////////////////////////////

declare const s: Select<"A", { l: (k: number) => Close }>;

// @dts-jest:pass:snap Close
send<"A", { l: (k: number) => Close }, "l">(s, "A", "l", 42);

// @dts-jest:fail:snap Argument of type '"B"' is not assignable to parameter of type '"A"'.
send<"A", { l: (k: number) => Close }, "l">(s, "B", "l", 42);

// @dts-jest:fail:snap Type '"A"' is not assignable to type '"B"'.
send<"B", { l: (k: number) => Close }, "l">(s, "A", "l", 42);

// @dts-jest:fail:snap Argument of type '"r"' is not assignable to parameter of type '"l"'.
send<"A", { l: (k: number) => Close }, "l">(s, "A", "r", 42);

// @dts-jest:fail:snap Type '"r"' does not satisfy the constraint '"l"'.
send<"A", { l: (k: number) => Close }, "r">(s, "A", "l", 42);

// @dts-jest:fail:snap Property 'r' is missing in type '{ l: (k: number) => Close; }'.
send<"A", { r: (k: number) => Close }, "r">(s, "A", "l", 42);

// @dts-jest:fail:snap Argument of type 'string' is not assignable to parameter of type 'number'.
send<"A", { l: (k: number) => Close }, "l">(s, "A", "l", "42");

// @dts-jest:fail:snap Type 'string' is not assignable to type 'number'.
send<"A", { l: (k: string) => Close }, "l">(s, "A", "l", 42);

//////////////////////////// test cases of receive ////////////////////////////

declare const o: Offer<"B", ["l", [number, Close]]>;

// @dts-jest:pass:snap Promise<{ label: "l"; value: number; port: Close }>
receive<"B", ["l", [number, Close]]>(o, "B");

// @dts-jest:fail:snap Argument of type '"A"' is not assignable to parameter of type '"B"'.
receive<"B", ["l", [number, Close]]>(o, "A");

// @dts-jest:fail:snap Type '"B"' is not assignable to type '"A"'.
receive<"A", ["l", [number, Close]]>(o, "B");

// @dts-jest:fail:snap Type '"l"' is not assignable to type '"r"'.
receive<"B", ["r", [number, Close]]>(o, "B");

// @dts-jest:fail:snap Type 'number' is not assignable to type 'string'.
receive<"B", ["l", [string, Close]]>(o, "B");
