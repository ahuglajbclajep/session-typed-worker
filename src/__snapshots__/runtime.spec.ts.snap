// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Argument of type '"A"' is not assignable to parameter of type '"B"'. (type) should match snapshot 1`] = `"Argument of type '\\"A\\"' is not assignable to parameter of type '\\"B\\"'."`;

exports[`Argument of type '"B"' is not assignable to parameter of type '"A"'. (type) should match snapshot 1`] = `"Argument of type '\\"B\\"' is not assignable to parameter of type '\\"A\\"'."`;

exports[`Argument of type '"r"' is not assignable to parameter of type '"l"'. (type) should match snapshot 1`] = `"Argument of type '\\"r\\"' is not assignable to parameter of type '\\"l\\"'."`;

exports[`Argument of type 'string' is not assignable to parameter of type 'number'. (type) should match snapshot 1`] = `"Argument of type 'string' is not assignable to parameter of type 'number'."`;

exports[`Close (type) should match snapshot 1`] = `"any"`;

exports[`Promise<["l", [number, Close]]> (type) should match snapshot 1`] = `"any"`;

exports[`Property 'r' is missing in type '{ l: (k: number) => Close; }'. (type) should match snapshot 1`] = `
"Argument of type 'Select<\\"A\\", { l: (k: number) => Close; }>' is not assignable to parameter of type 'Select<\\"A\\", { r: (k: number) => Close; }>'.
  Property 'r' is missing in type '{ l: (k: number) => Close; }' but required in type '{ r: (k: number) => Close; }'."
`;

exports[`Type '"A"' is not assignable to type '"B"'. (type) should match snapshot 1`] = `
"Argument of type 'Select<\\"A\\", { l: (k: number) => Close; }>' is not assignable to parameter of type 'Select<\\"B\\", { l: (k: number) => Close; }>'.
  Type '\\"A\\"' is not assignable to type '\\"B\\"'."
`;

exports[`Type '"B"' is not assignable to type '"A"'. (type) should match snapshot 1`] = `
"Argument of type 'Offer<\\"B\\", [\\"l\\", [number, Close]]>' is not assignable to parameter of type 'Offer<\\"A\\", [\\"l\\", [number, Close]]>'.
  Type '\\"B\\"' is not assignable to type '\\"A\\"'."
`;

exports[`Type '"l"' is not assignable to type '"r"'. (type) should match snapshot 1`] = `
"Argument of type 'Offer<\\"B\\", [\\"l\\", [number, Close]]>' is not assignable to parameter of type 'Offer<\\"B\\", [\\"r\\", [number, Close]]>'.
  Type '[\\"l\\", [number, Close]]' is not assignable to type '[\\"r\\", [number, Close]]'.
    Type '\\"l\\"' is not assignable to type '\\"r\\"'."
`;

exports[`Type '"r"' does not satisfy the constraint '"l"'. (type) should match snapshot 1`] = `"Type '\\"r\\"' does not satisfy the constraint '\\"l\\"'."`;

exports[`Type 'number' is not assignable to type 'string'. (type) should match snapshot 1`] = `
"Argument of type 'Offer<\\"B\\", [\\"l\\", [number, Close]]>' is not assignable to parameter of type 'Offer<\\"B\\", [\\"l\\", [string, Close]]>'.
  Type '[\\"l\\", [number, Close]]' is not assignable to type '[\\"l\\", [string, Close]]'.
    Type '[number, Close]' is not assignable to type '[string, Close]'.
      Type 'number' is not assignable to type 'string'."
`;

exports[`Type 'string' is not assignable to type 'number'. (type) should match snapshot 1`] = `
"Argument of type 'Select<\\"A\\", { l: (k: number) => Close; }>' is not assignable to parameter of type 'Select<\\"A\\", { l: (k: string) => Close; }>'.
  Type '{ l: (k: number) => Close; }' is not assignable to type '{ l: (k: string) => Close; }'.
    Types of property 'l' are incompatible.
      Type '(k: number) => Close' is not assignable to type '(k: string) => Close'.
        Types of parameters 'k' and 'k' are incompatible.
          Type 'string' is not assignable to type 'number'."
`;
